src/lib.rs

```diff
@@ -324,9 +324,30 @@ pub fn optimize(input: &InFile, output: &OutFile, opts: &Options) -> PngResult<(
 
     let deadline = Arc::new(Deadline::new(opts.timeout));
 
+    // grab metadata before even opening input file to preserve atime
+    let opt_metadata_preserved;
     let in_data = match *input {
-        InFile::Path(ref input_path) => PngData::read_file(input_path)?,
+        InFile::Path(ref input_path) => {
+            if opts.preserve_attrs {
+                opt_metadata_preserved = input_path
+                    .metadata()
+                    .map_err(|err| {
+                        // TODO: Fail if input and output file are the same and metadata cannot be preserved?
```

Comment:
This is a very good question. My gut tells me that we should fail if the user wants to preserve metadata and they are unable to. Then the user can remove the preserve metadata flag if this is acceptable.

Reply:
implemented the more rigorous failure mode in 0001346, please review :)